### PR TITLE
TPU gradient reversal layer

### PR DESCRIPTION
**Pull request by dizcology**
Customer asked whether the [gradient reversal layer](https://arxiv.org/abs/1505.07818) can be used with TPU training.  A working prototype is implemented here.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/cloudml-samples/301)

### CODE REVIEW

File: tpu/templates/tpu_grl_estimator/trainer.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import numpy as np
+import tensorflow as tf
+from tensorflow.contrib import summary
```
Comment (gogasca) [inline tpu/templates/tpu_grl_estimator/trainer.py]: is there an alternative not in contrib ?

File: tpu/templates/tpu_grl_estimator/trainer.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+#
+# Both the label predictor and the domain classifier will try to minimize 
+# classification loss, but the gradients backpropagated from the domain
+# classifier to the feature extractor have their signs reversed.
+#
+#
+# This model function also shows how to use `host_call` to output summaries.
+#
+
+def model_fn(features, labels, mode, params):
```
Comment (gogasca) [inline tpu/templates/tpu_grl_estimator/trainer.py]: can we use params as args using name space: params.gr_weight vs params['gr_weight']

File: tpu/templates/tpu_grl_estimator/trainer.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+        if 'COLAB_TPU_ADDR' in os.environ:
+            tpu_grpc = 'grpc://{}'.format(os.environ['COLAB_TPU_ADDR'])
+
+            args.tpu = tpu_grpc
+            args.use_tpu = True
+
+            # Upload credentials to the TPU
+            with tf.Session(tpu_grpc) as sess:
+                data = json.load(open('/content/adc.json'))
+                tf.contrib.cloud.configure_gcs(sess, credentials=data)
```
Comment (gogasca) [inline tpu/templates/tpu_grl_estimator/trainer.py]: Is there an alternative to tf.contrib.cloud.configure_gcs

File: tpu/templates/tpu_grl_estimator/trainer.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+            config=config,
+            params=params)
+
+    estimator.train(train_input_fn, max_steps=args.max_steps)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
```
Comment (gogasca) [inline tpu/templates/tpu_grl_estimator/trainer.py]: Should we add help to args, explaining the meaning of args, (Even though some are self-explanatory)

File: tpu/templates/tpu_grl_estimator/trainer.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+        help='The weight for gradient reversal.')
+    parser.add_argument(
+        '--lambda',
+        default=1.0,
+        help='The regularization factor.')
+
+    args, _ = parser.parse_known_args()
+
+    # colab.research.google.com specific
+    import sys
```
Comment (gogasca) [inline tpu/templates/tpu_grl_estimator/trainer.py]: imports should be in top. https://www.python.org/dev/peps/pep-0008/#imports

File: tpu/templates/tpu_grl_estimator/trainer.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+            save_checkpoints_steps=args.save_checkpoints_steps,
+            save_summary_steps=100)
+
+        # TPUEstimator
+        estimator = tf.contrib.tpu.TPUEstimator(
+            model_fn=model_fn,
+            config=config,
+            params=params,
+            train_batch_size=args.train_batch_size,
+            eval_batch_size=32, # FIXME
```
Comment (gogasca) [inline tpu/templates/tpu_grl_estimator/trainer.py]: What does # FIXME means?

File: tpu/templates/tpu_grl_estimator/trainer.py (hunk 2 above)
Comment (dizcology) [reply tpu/templates/tpu_grl_estimator/trainer.py]: I think `params` needs be a dict, as the call to build the estimator sometimes injects/updates additional keys/values to it.

File: tpu/templates/tpu_grl_estimator/trainer.py (hunk 3 above)
Comment (dizcology) [reply tpu/templates/tpu_grl_estimator/trainer.py]: let me look into that as a follow up PR around the time when TF 2.0 is releases.

File: tpu/templates/tpu_grl_estimator/trainer.py (hunk 5 above)
Comment (dizcology) [reply tpu/templates/tpu_grl_estimator/trainer.py]: I will try to fix this for the followup PR to remove colab specific codes from the .py file.

File: tpu/templates/tpu_grl_estimator/trainer.py (hunk 6 above)
Comment (dizcology) [reply tpu/templates/tpu_grl_estimator/trainer.py]: remnant from experimental code, thanks for catching it.  i see that it also exists in other tpu templates - so i'll fix this in the next cross sample PR.

**Comment by dizcology** [comment]
thanks @gogasca 

Several issues affect multiple templates in the cloudml-samples/tpu/templates directory.  I'll create a separate PR to fix them at once.  I created this issue to track them: https://github.com/GoogleCloudPlatform/cloudml-samples/issues/326

**Pull request merged**